### ISSUE TITLE
investigate cache headers on /sign_in

### ISSUE THREAD

**Issue by jrgm** [mozilla/persona#3547]
There were reports of people getting 404 for content from static.login.persona.org, such as me, jedp and ednapirana. 

ednapirana's url was https://static.login.persona.org/v/f915b0aa15/production/en/dialog.css. 

That would be a correct url on the train-2013.04.26 release. (Production was updated to train-2013.05.22 at 2:30pm PDT today). 

```
curl -k -s https://persona-org-0525-1045101880.us-west-2.elb.amazonaws.com/v/f915b0aa15/production/en/dialog.css | md5sum
f915b0aa1558424be23575603f082c32  -
```

But the /sign_in dialog is served with `Cache-Control: public, max-age=0` and an ETag header, so a browser should request again each time the dialog comes up with 'If-None-Match', and that's what I see it doing with FF21. 

Anyways, is it a problem with the etags, or some weird browser cacheing, or weird DNS, or...

**Comment by jrgm** [comment]
Perhaps we should add 'must-revalidate' to 'cache-control', since we really want the GET under all circumstances. 

Mark Nottingham in http://www.mnot.net/cache_docs/

```
must-revalidate — tells caches that they must obey any freshness information you give them about a 
representation. HTTP allows caches to serve stale representations under special conditions; by specifying this 
header, you’re telling the cache that you want it to strictly follow your rules.
```

Relevant edge case (but have no data that shows this applies to the problems we saw, but...) http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.8 second paragraph.

**Comment by jrgm** [comment]
mnot's redbot elaborates some more - http://redbot.org/?uri=https%3A%2F%2Flogin.persona.org%2Fsign_in

(Still I don't know if unreachable or 5xx applies here without log data, but...)

**Comment by jaredhirsch** [comment]
I think we could add proxy-revalidate and must-revalidate to cache-control.

We could also add an Expires header in the past for proxies/caches that only speak HTTP 1.0.

**Comment by jaredhirsch** [comment]
After looking at a whole lot of docs on the subject, going with @jrgm's original suggestion for now. We'll see if this recurs after the next release.

HTTP caching implementations are a mess

**Comment by jaredhirsch** [comment]
**Update:** TL;DR this actually appears to be a Firefox browser bug.

The issue recurred with today's production release, and @jbuck was kind enough to help us debug in detail. He confirmed that @ednapiranha was also seeing this problem again.

The issue (at least, today) was not stale copies of sign_in referring to gone static endpoints--the page was fresh, but the static files (with the correct cachify-generated path) were 404ing.

It looks like firefox caches DNS records and reuses them, even if their TTL has expired. He was able to resolve the problem by clearing firefox's DNS cache, adding this pref to about:config: `network.dnsCacheExpiration = 0`. This forced FF to fall back to the DNS cache of the underlying OS, which was correctly resolving to the new AWS stack according to output from `dig`.

@jrgm dug up a bugzilla bug from a few months back https://bugzilla.mozilla.org/show_bug.cgi?id=861273 which points out that stale DNS entries can linger for up to a month.

:beers: all around

**Comment by jaredhirsch** [comment]
Adding a bit more context: FF 18 made DNS lookups asynchronous. Here's the description of the network.dnsCacheExpirationGracePeriod pref from [moz-central](http://dxr.allizom.org/mozilla-central/source/modules/libpref/src/init/all.js#l1256):

> ```
> // The grace period allows the DNS cache to use expired entries, while kicking off
> // a revalidation in the background. In seconds, but rounded to minutes in gecko.
> // Default to 30 days. (basically forever)
> pref("network.dnsCacheExpirationGracePeriod", 2592000);
> ```

This makes it pretty clear what's happening: FF caches DNS lookups indefinitely, and refreshes that cache async. This doesn't, however, explain why only a few users seem to be affected.

**Comment by jrgm** [comment]
So, I've been running a loop to load https://static.login.persona.org/v/a8b2c60b06/production/en/browserid.js in FF22. At the same time, I'm looking at NSPR logging for nsHostResolver:5 and also watching tcpdump for port 53 (on OSX).

After startup, according to FF's own NSPR logs, and from tcpdump, FF never re-resolves the IP for static.login.persona.org. However, then I used `ipfw` to deny outbound port 443. That, of course, means a failure for the next attempt to open that URL in FF. Then I removed the `ipfw` deny 443 rule. Next attempt to load the URL works again, but the logging shows that a different code path is being used, in particular that it now logs `Using positive cache entry for host [static.login.persona.org] but starting async renewal.` on each attempt, and tcpdump shows that new DNS requests are happening. Anyways, this seems buggy.

However, our problem is really "what do we do next time we push a release", and also "are other browsers really ok?". 

So, this is going a bit outside the original scope of this issue, but things to look into:
- I will talk to a couple of network engineers here in MV tomorrow to see if they are already aware of this behaviour with FF (or if I'm misinterpreting things).
- Perhaps, when we cut over DNS to the new stacks, we brutally stop taking traffic on the old stacks. (This will not be pretty, but may get FF users, after the first failure, back into async checking the right IPs). Obviously I don't like this option.
- Investigate if there is a way with ELB, EIP, Route53, etc. to always map *.login.persona.org to the same set of IPs. (Hand waving here).
- Get the access logs from the stack-0613 EC2 and analyze which browsers were not re-resolving DNS (i.e., was it only FF?)
- Consider setting `no-store, no-cache, no-im-not-kidding` `Cache-Control` options for `/sign_in`
- Other?
